packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+        return;
+      }
+
+      // Template literal index signatures count as used if any remaining property's type is the
+      // same, this may miss some unused template literal index signatures
+      if (
+        tsutils.isTemplateLiteralType(indexParameterType) &&
+        indexSignatureHasMatchingValue(remainingProperties, member)
+      ) {
+        return;
+      }
+
+      reportOnMember(member, {
+        type: 'index signature',
+        key: `[${checker.typeToString(indexParameterType)}]`,
+      });
+    }
+
+    /**
+     * Checks if an non index-signature property is covered by at least one destructure property.
+     */
+    function checkStaticMember(
+      member: TSESTree.TSMethodSignature | TSESTree.TSPropertySignature,
+      remainingProperties: RemainingProperties,
+      dynamicProperties: DynamicProperties,
+    ): void {
+      const memberKey = getStaticMemberAccessValue(member, context);
+
+      if (memberKey === undefined) {
+        return;
+      }
+
+      // check if this is used by a remaining property
+      const remainingProperty = remainingProperties.get(memberKey);
+
+      if (remainingProperty) {
+        remainingProperties.delete(memberKey);
+
+        if (
+          member.type === AST_NODE_TYPES.TSPropertySignature &&
+          member.typeAnnotation
+        ) {
+          checkParam(
+            remainingProperty.property.value,
+            member.typeAnnotation.typeAnnotation,
+          );
+        }
+
+        return;
+      }
+
+      // check if this is used by a dynamic index type
+      const dynamicProperty = getDynamicKeyForMember(
+        member,
+        memberKey,
+        dynamicProperties,
+      );
+
+      if (dynamicProperty) {
+        if (
+          member.type === AST_NODE_TYPES.TSPropertySignature &&
+          member.typeAnnotation
+        ) {
+          checkParam(
+            dynamicProperty.value,
+            member.typeAnnotation.typeAnnotation,
+          );
+        }
+
+        return;
+      }
+
+      reportOnMember(member, {
+        type: 'property',
+        key: String(memberKey),
+      });
+    }
+
+    /**
+     * Attempts to find a dynamic property that matches a given member key.
+     */
+    function getDynamicKeyForMember(
+      member: TSESTree.TSMethodSignature | TSESTree.TSPropertySignature,
+      memberKey: number | string | symbol,
+      dynamicProperties: DynamicProperties,
+    ): TSESTree.Property | undefined {
+      for (const destructure of dynamicProperties) {
+        destructure.type ??= getConstrainedTypeAtLocation(
+          services,
+          destructure.property.key,
+        );
+
+        for (const type of tsutils.unionTypeParts(destructure.type)) {
+          if (tsutils.isStringLiteralType(type) && type.value === memberKey) {
+            return destructure.property;
+          }
+
+          if (tsutils.isNumberLiteralType(type) && type.value === memberKey) {
+            return destructure.property;
+          }
```

Comment:
[Refactor] This could be condensed into a more dense `if((string || number) && ===`, or even better, switched to the more general API for less code:

```suggestion
          if (tsutils.isLiteralType(type) && type.value === memberKey) {
            return destructure.property;
          }
```